### PR TITLE
feat(ec2): peer cidr ip validation

### LINKED ISSUE #3639: EC2: Security Group Peer CIDR validation
**Note: for support questions, please first reference our [documentation](https://docs.aws.amazon.com/cdk/api/latest), then use [Stackoverflow](https://stackoverflow.com/questions/ask?tags=aws-cdk)**. This repository's issues are intended for feature requests and bug reports.

* **I'm submitting a ...**
  - [ ] :beetle: bug report
  - [x] :rocket: feature request
  - [ ] :books: construct library gap
  - [ ] :phone: security issue or vulnerability => Please see [policy](https://github.com/aws/aws-cdk/security/policy)
  - [ ] :question: support request => Please see note at the top of this template.

* **What is the current behavior?**
*If the current behavior is a :beetle:bug:beetle:: Please provide the steps to reproduce*
`Peer.ipv4` and `Peer.ipv6` do not throw when given an invalid CIDR block. The error only happens when deploying the stack.

* **What is the expected behavior (or behavior of feature suggested)?**
The parameter should be regex matched, and an error should be thrown if the provided parameter is not a valid CIDR. Ideally, the CDK could also match a plain IP without the mask, and throw a different error, suggesting adding it, e.g. `did you mean 0.0.0.0/0, or 255.255.255.255/24?`.

* **What is the motivation / use case for changing the behavior or adding this feature?**
The naming of the `Peer.ipv4/v6` methods is misleading, and I keep forgetting that the parameter must be a CIDR.

* **Please tell us about your environment:**

  - CDK CLI Version: 1.3.0
  - Module Version: 1.3.0
  - OS: all
  - Language: all

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. associated pull-request, stackoverflow, gitter, etc)

### PR DESCRIPTION
**Pull request by nmussy**
Fixes #3639

----

Please read the [contribution guidelines](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md) and follow the pull-request checklist.

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-ec2/lib/peer.ts [hunk 1]
```diff
@@ -112,6 +122,16 @@ class CidrIPv6 implements IPeer {
   public readonly uniqueId: string;
 
   constructor(private readonly cidrIpv6: string) {
+    const cidrMatch = cidrIpv6.match(/^([\da-f]{0,4}:){2,7}([\da-f]{0,4})?(\/\d+)?$/);
```
Comment (nmussy) [inline packages/@aws-cdk/aws-ec2/lib/peer.ts]: I used some fairly loose RegExps, especially for IPv6. The stricter ones tend to become a little crazy

File: packages/@aws-cdk/aws-ec2/lib/peer.ts [hunk 2]
```diff
@@ -77,6 +77,16 @@ class CidrIPv4 implements IPeer {
...
 
   constructor(private readonly cidrIp: string) {
+    const cidrMatch = cidrIp.match(/^(\d{1,3}\.){3}\d{1,3}(\/\d+)?$/);
+
+    if (!cidrMatch) {
+      throw new Error(`Invalid IPv4 CIDR: "${cidrIp}"`);
+    }
+
+    if (!cidrMatch[2]) {
+      throw new Error(`CIDR mask is missing in IPv4: "${cidrIp}". Did you mean "${cidrIp}/xx"?`);
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/lib/peer.ts]: I feel it makes more sense to replace `/xx` with `/32` in the error message. If people wrote a single IP address, that's probably what they meant, and if they didn't it's clear enough where to substitute values to make it do what they want.

I considered automatically adding `/32` to the IP address, but we should probably make a separate class for single IP addresses.

File: packages/@aws-cdk/aws-ec2/lib/peer.ts [hunk 3]
```diff
@@ -77,6 +77,16 @@ class CidrIPv4 implements IPeer {
   public readonly uniqueId: string;
 
   constructor(private readonly cidrIp: string) {
+    const cidrMatch = cidrIp.match(/^(\d{1,3}\.){3}\d{1,3}(\/\d+)?$/);
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/lib/peer.ts]: I guess to be safe this needs a `if (!cdk.Token.isUnresolved(cidrIp)) { ... }` block around the checks, same for the IPv6 version.

File: packages/@aws-cdk/aws-ec2/lib/peer.ts [hunk 4]
```diff
@@ -112,6 +122,16 @@ class CidrIPv6 implements IPeer {
...
 
   constructor(private readonly cidrIpv6: string) {
+    const cidrMatch = cidrIpv6.match(/^([\da-f]{0,4}:){2,7}([\da-f]{0,4})?(\/\d+)?$/);
+
+    if (!cidrMatch) {
+      throw new Error(`Invalid IPv6 CIDR: "${cidrIpv6}"`);
+    }
+
+    if (!cidrMatch[3]) {
+      throw new Error(`CIDR mask is missing in IPv6: "${cidrIpv6}". Did you mean "${cidrIpv6}/xx"?`);
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/lib/peer.ts]: `/128` I suppose?

File: packages/@aws-cdk/aws-ec2/lib/peer.ts (hunk 2 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-ec2/lib/peer.ts]: > we should probably make a separate class for single IP addresses.

I like that idea. We could also deprecate `ipv4/v6` and explicitly name them `ipv4Cidr`

File: packages/@aws-cdk/aws-ec2/lib/peer.ts (hunk 4 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-ec2/lib/peer.ts]: Done

File: packages/@aws-cdk/aws-ec2/lib/peer.ts (hunk 2 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-ec2/lib/peer.ts]: Otherwise, done

File: packages/@aws-cdk/aws-ec2/lib/peer.ts (hunk 3 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-ec2/lib/peer.ts]: Done

**Pull request merged**